Add support for calling C++ constructors

martinboehme opened the pull request:
A problem that we have to deal with when calling C++ constructors is that the ABI of a C++ constructor depends not only on the signature of the constructor but on the class on which it is defined. For example, it can be necessary to pass additional “implicit” arguments to constructors of virtual base classes.

It is therefore fundamentally not possible for SignatureExpansion (which doesn’t have access to the `CXXConstructorDecl`) to reliably produce the correct LLVM signature from only the type of the constructor. The approach we take is to try to get most of the way in SignatureExpansion by marking the SIL function return type with the `@out` attribute, representing the fact that C++ constructors take a `this` points to the object to be initialized.

This produces an “assumed” constructor ABI in SignatureExpansion that assumes there are no implicit arguments and that the return type of the constructor ABI is void (and indeed there is no way to represent anything else in the SIL type). If this assumed ABI doesn’t match the actual ABI, we insert a thunk in IRGen.

The thunk is marked `alwaysinline`, so it doesn’t incur any runtime overhead, but some compile-time overhead is required for LLVM to inline the thunk. If this turns out to be excessive, we can introduce what is essentially a peephole optimization for a full apply of a `function_ref`that refers to a C++ constructor and directly emit the constructor call in this case (along with any required implicit arguments) instead of calling the thunk. I’ve chosen not to tackle that in this PR though as it’s already large enough.

On some ABIs (e.g. Itanium x64), we get lucky and the ABI for a complete constructor call always matches the assumed ABI.

In addition to the core C++ constructor functionality, this PR adds some a couple of small required features:

*  Conversion of thin to a Clang `void` type
*  Handling of formal indirect results in `SignatureExpansion::expandExternalSignatureTypes()`

martinboehme committed:
Add support for calling C++ constructors.

Because C++ constructors always take a `this` pointer to the object to
be initialized, we mark the SIL function return type with the `@out`
attribute.

On the IRGen side, we retrofit support for formal indirect return values as
well as thin metatypes.

varungandhi-apple commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3804,26 +3835,44 @@ namespace {
+
+      AbstractFunctionDecl *result = nullptr;
+      if (auto *ctordecl = dyn_cast<clang::CXXConstructorDecl>(decl)) {
+        // TODO: Is failable, throws etc. correct?
+        DeclName ctorName(Impl.SwiftContext, DeclBaseName::createConstructor(),
+                          bodyParams);
+        result = Impl.createDeclWithClangNode<ConstructorDecl>(
+            decl, AccessLevel::Public, ctorName, loc, /*failable=*/false,
+            /*FailabilityLoc=*/SourceLoc(), /*Throws=*/false,
+            /*ThrowsLoc=*/SourceLoc(), bodyParams, /*GenericParams=*/nullptr,
+            dc);
```
Does this mean the Clang Importer will incorrectly translate constructors that throw on failure to non-throwing constructors? That doesn't sound right... I can understand if you want to punt it to a later time though, but please file a JIRA (and link that here) if you are doing that.

compnerd replied:
I think that adding a test case for the problem would also be a good idea.

martinboehme replied:
> Does this mean the Clang Importer will incorrectly translate constructors that throw on failure to non-throwing constructors? That doesn't sound right...

Agree, this definitely shouldn't happen. I plan to resolve the `TODO` one way or the other before this PR is ready for review (I've put the PR up here for the time being to get feedback on the general approach).

I think what I'll probably do initially is to only import C++ constructors that are marked `noexcept`. Importing throwing C++ constructors (and mapping them to throwing Swift initializers) would then come as a second step. In this case, I assume filing a Jira ticket would not be necessary in this case (as we are then talking only about missing functionality, not erroneous functionality)?

And yes, I will definitely add test cases for this.

martinboehme replied:
I've added a commit (with tests) that restricts imports only to `noexcept` constructors for the time being.

martinboehme replied:
I've thought about exception handling some more and have written up a separate post for it:

https://forums.swift.org/t/handling-c-exceptions/34823/2

tl;dr: I don't think potentially-throwing C++ constructors (which really just means C++ constructors not marked `noexcept`) should be mapped to `throws` initializers because it results in non-idiomatic C++ code. Many C++ constructors don't in fact throw, but aren't explicitly marked `noexcept`; the same is true for functions, and indeed, we currently import potentially-throwing C++ functions as non-throwing Swift functions. Instead, the default behavior should be to import C++ functions as non-throwing; in case the C++ function does throw, we terminate the program.

varungandhi-apple replied:
>  In this case, I assume filing a Jira ticket would not be necessary (as we are then talking only about missing functionality, not erroneous functionality)?

Yes, that's correct. Thanks for starting the broader discussion.

martinboehme replied:
As per the discussion on the forum, we want to import all C++ constructors, whether they are marked `noexcept` or not, as non-throwing initializers. I've therefore reverted the change that only imported `noexcept` constructors.

martinboehme committed:
Only import constructors marked `noexcept`.

varungandhi-apple commented on lib/IRGen/GenClangType.cpp:
```diff
@@ -403,7 +403,19 @@ clang::CanQualType GenClangType::visitProtocolType(CanProtocolType type) {
 
 clang::CanQualType GenClangType::visitMetatypeType(CanMetatypeType type) {
-  return getClangMetatypeType(getClangASTContext());
+  assert(type->hasRepresentation() &&
+         "metatype should have been assigned a representation by SIL");
+  switch (type->getRepresentation()) {
+  case MetatypeRepresentation::Thin:
+    return getClangASTContext().VoidTy;
+
+  case MetatypeRepresentation::Thick:
+    llvm_unreachable("thick metatypes don't have a corresponding Clang type");
+
+  case MetatypeRepresentation::ObjC:
+    return getClangMetatypeType(getClangASTContext());
+  }
+  llvm_unreachable("bad representation");
```
For the changes in this file, please also update the corresponding code in `ClangTypeConverter.cpp`.

martinboehme replied:
Thanks for pointing this out -- I hadn't seen the comment at the top of `IRGen/GenClangType.cpp` saying to do this.

Two questions:

* In IRGen, I can rely on the metatype's representation being populated (see assertion that I added above, copied from `TypeConverter::convertMetatypeType()`). But can I also rely on the representation being populated in `AST/ClangTypeConverter.cpp`? If not, what should I do -- check whether we have a representation, and if we don't, just make an assumption...?

* Any insights into why there's this duplication between `AST/ClangTypeConverter.cpp` and `IRGen/GenClangType.cpp`? The main difference between them seems to be that the latter computes canonical types while the latter doesn't, but that could presumably be dealt with via a wrapper?

varungandhi-apple replied:
> In IRGen, I can rely on the metatype's representation being populated (see assertion that I added above, copied from TypeConverter::convertMetatypeType()). But can I also rely on the representation being populated in AST/ClangTypeConverter.cpp?

I'm not 100% sure but I think you should be able to rely on it. If you see any failure purely because of this change, you can file a bug report on JIRA for investigation (and assign it to me) leaving a placeholder branch here to avoid a crash, say something like:

```
// FIXME: SR-NNNNN Unclear if we should expect to have a representation here.
if (!type->hasRepresentation()) {
    return getClangIdType(clangCtx);
}
```

Since the results of `ClangTypeConverter.cpp` are not really being used per se, it's ok if we have that here as a stop-gap until we start using it.

> Any insights into why there's this duplication between AST/ClangTypeConverter.cpp and IRGen/GenClangType.cpp? The main difference between them seems to be that the latter computes canonical types while the latter doesn't, but that could presumably be dealt with via a wrapper?

Going forward, we want to be able to track Clang types in the AST and SIL. (You'll see that `FunctionType` has a field for it and I have an open PR for SIL which I haven't had time to work on recently). The plan is to eventually get rid of the IRGen one and use the AST one.

There are some subtle differences the two implementations are not exactly the same, but I _think_ it shouldn't matter for this change.

martinboehme replied:
OK, thanks. Done!

(Finally -- sorry, I was busy with other things and it took me a while to get back to this.)

martinboehme committed:
Revert "Only import constructors marked `noexcept`."

This reverts commit 29650d8c1f302708a32304d49c703c9ddbf30b75.

As discussed here, we want to import all constructors (whether they are
marked `noexcept` or not) as non-throwing initializers;

https://forums.swift.org/t/handling-c-exceptions/34823/50

martinboehme committed:
Add a constructor thunk if required to add additional constructor
arguments.

Also add more IR tests and make various other changes.

martinboehme committed:
Move tests from `CXXInterop` to `Interop/Cxx`.

This is the canonical location that we've agreed on.

martinboehme committed:
Duplicate changes to GenClangType in ClangTypeConverter.

See discussion here:
https://github.com/apple/swift/pull/30630#discussion_r398967412

MForster commented on lib/ClangImporter/ImportName.cpp:
```diff
@@ -1407,6 +1407,20 @@ ImportedName NameImporter::importNameImpl(const clang::NamedDecl *D,
+    isFunction = true;
+    result.info.initKind = CtorInitializerKind::Designated;
+    baseName = "init";
+    // Add empty argument names.
+    if (auto ctor = dyn_cast<clang::CXXConstructorDecl>(D)) {
+      for (size_t i = 0; i < ctor->param_size(); ++i) {
+        argumentNames.push_back(StringRef());
+      }
+      if (ctor->isVariadic())
+        argumentNames.push_back(StringRef());
+    }
```
Maybe don't duplicate this code. This is identical for functions and it will be the same for operators.

martinboehme replied:
Pulled it out into a separate function. This has also allowed me to eliminate the comment, as it now just duplicates what the code says anyway.

MForster commented on test/Interop/Cxx/class/Inputs/module.modulemap:
```diff
@@ -2,6 +2,10 @@ module AccessSpecifiers {
   header "access-specifiers.h"
 }
 
+module CxxConstructors {
+  header "cxx-constructors.h"
+}
```
The `cxx-` prefix is redundant. This is in a directory that is already called `Cxx`.

martinboehme replied:
Good point -- removed the prefix (and also in all the test names).

martinboehme committed:
Remove redundant "cxx" from test names.

martinboehme committed:
Eliminate duplication of code for adding empty argument names.

martinboehme committed:
Various changes after merging master:

- Adapt tests to changes that have happened in the meantime (e.g.
  `HasVirtualBase` is rightly no longer considered loadable)
- Don't import copy or move constructors (noticed this because
  references are now imported correctly, so copy and move constructors
  suddenly started showing up in the SIL test)
- Don't try to define an implicitly-deleted default constructor (this
  previously broke loadable-types-silgen.swift)

martinboehme commented:
@swift-ci Please test

martinboehme commented:
@swift-ci Please test Windows

martinboehme commented:
This is now ready for review again -- sorry for the delay.

See also the discussion here:

https://forums.swift.org/t/calling-c-constructors-looking-for-feedback-on-my-approach/34787/18?u=mboehme

martinboehme commented:
@rjmccall Not sure if you want to review this also?

gribozavr commented on test/Interop/Cxx/class/constructors-ir.swift:
```diff
@@ -0,0 +1,42 @@
+// Target-specific tests for C++ constructor call code generation.
```
Please rename the file to `...-irgen`.

martinboehme replied:
Done.

gribozavr commented on test/Interop/Cxx/class/constructors-ir.swift:
```diff
@@ -0,0 +1,42 @@
+  // Note `this` return type and implicit "most derived" argument.
+  // MICROSOFT_X64: call %struct.HasVirtualBase* @"??0HasVirtualBase@@QEAA@UArgType@@@Z"(%struct.HasVirtualBase* %{{[0-9]+}}, i32 %{{[0-9]+}}, i32 1)
+  return HasVirtualBase(ArgType())
+}
```
Could you add some more tests, for cases that don't involve virtual base classes?

For example, you mentioned that constructors on some Itanium ABI variants on ARM64 iOS return 'this', so they would also need a thunk.

Also would be nice to have some tests for cases that don't need thunks.

Feel free to add some of them here if you are not using standard library types -- or in a separate file if you need standard library types.

martinboehme replied:
> Could you add some more tests, for cases that don't involve virtual base classes?

Done.

> For example, you mentioned that constructors on some Itanium ABI variants on ARM64 iOS return 'this', so they would also need a thunk.

Some of the existing test cases already demonstrate this -- see the comments on the `ITANIUM_ARM` and `MICROSOFT_X64` tests, which point out the `this` return type.

> Also would be nice to have some tests for cases that don't need thunks.

This was already the case for the existing `ITANIUM_X64` test. The newly added `ITANIUM_X64` test also does not need a thunk.

gribozavr commented on test/Interop/Cxx/class/constructors-ir.swift:
```diff
@@ -0,0 +1,42 @@
+  //   attributes of the constructor that was generated by Clang, which doesn't
+  //   insert these attributes.
+  //
+  // - The `this` parameter should _not_ carry a `nocapture` attribute (unlike
```
If the type is loadable in Swift, `nocapture` is permitted, I think.

In this example, the function returns the result indirectly, and the result is marked `nocapture`. Since we construct a C++ object in that memory, the constructor must also be `nocapture`

martinboehme replied:
Hm... good point.

Since there is no way in general to guarantee that the C++ constructor will be `nocapture`, I think the conclusion here is that when we're returning C++ objects indirectly, we should always do so without `nocapture`?

(We could, in principle, add the `nocapture` back if we can prove to ourselves that the pointer can't be captured. But I think Swift currently isn't set up to do that -- I assume the place that adds the `nocapture` is in `SignatureExpansion`, and that only sees the type of the function, not its definition.)

rjmccall replied:
Well, we can probably check whether the type is trivially copyable.  If it's trivially copyable, it can't be captured, because Swift is absolutely not going to promise not to move it around.

martinboehme replied:
I think I see what you're saying. Swift can certainly also move non-trivially-copyable C++ types around as it pleases, but then the move constructor will notice this happening and can adjust the pointer that was previously captured by the constructor. But a trivially copyable type has no way of doing this, and so we can assume that the constructor doesn't capture `this`. Is this what you're getting at?

martinboehme replied:
FWIW, I've checked, and Clang doesn't add `nocapture` to the `this` parameter of a constructor, even if the type is trivially copyable. I'm not sure why this is, but if Clang isn't adding the `nocapture`, I don't think we should be adding it either.

There's a wider discussion of what we should be doing WRT to `nocapture` for Swift functions that return C++ types (see https://bugs.swift.org/browse/SR-12845), but that's outside the scope of this PR.

martinboehme committed:
Don't put an `sret` attribute on the `this` argument.

martinboehme committed:
Rename constructors-ir.swift to constructors-irgen.swift.

martinboehme committed:
Add additional IR tests for a class without a virtual base class.

martinboehme commented:
@swift-ci Please test

martinboehme commented:
@swift-ci Please test

martinboehme committed:
Add Windows-name-mangled version of symbol to SIL test.

martinboehme commented:
I've tried to reproduce the LLDB test failures locally, but they pass, even with ASAN enabled. Restarting testing in case this was just an infrastructure failure (though I'm not exactly sure how it could be).

martinboehme commented:
@swift-ci Please test

martinboehme committed:
Avoid crashing if lldb imports C++ structs without enabling C++ interop.

martinboehme commented:
@swift-ci Please test

martinboehme commented:
@swift-ci Please test Linux

martinboehme commented:
I've now fixed the LLDB failures, all tests are passing.

Ready for further review!

martinboehme committed:
Update comment in VisitCXXRecordDecl().

martinboehme commented:
@swift-ci Please test

martinboehme commented:
@swift-ci Please test Windows

lattner commented on lib/SIL/IR/SILFunctionType.cpp:
```diff
@@ -2708,6 +2708,25 @@ class CXXMethodConventions : public CFunctionTypeConventions {
     return ParameterConvention::Indirect_Inout;
   }
+  ResultConvention getResult(const TypeLowering &resultTL) const override {
+    if (dyn_cast<clang::CXXConstructorDecl>(TheDecl)) {
```
random drive by, but please use isa<> in a boolean context instead of dyn_cast.

There are also a number of 80 column violations that need to be fixed on comments and other things in this patch.

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -410,7 +410,19 @@ clang::QualType ClangTypeConverter::visitProtocolType(ProtocolType *type) {
+         "metatype should have been assigned a representation");
+  switch (type->getRepresentation()) {
+  case MetatypeRepresentation::Thin:
+    return ClangASTContext.VoidTy;
```
Does this work?  If the constructor takes a formal-in-SIL thing metatype argument, I'm surprised we can just translate it to `void`.

rjmccall replied:
Well, that's just saying that they're empty types; `TypeInfo` doesn't have a deeper meaning than that. But we can't just unconditionally turn a thin metatype into `void` if we're producing a foreign function signature; that's not legal in most positions.  Probably we need to erase the argument in the lowered type; I think we already have mechanics for that from import-as-member?

zoecarver replied:
@rjmccall I think you're right. We can't just turn these into `void`. I'm pretty sure that's what was causing the Windows bot to fail.

zoecarver replied:
I think the old implementation is actually sufficient (we'll see if it still works on a platform without objc-interop, though). The only time we should actually _use_ this path is when generating a lambda. If there's an issue with lambdas that take Meta types, I think that's a separate PR. This path is currently used when storing the converted clang type (and I think it's arguable whether we should actually be doing that at all).

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1117,6 +1129,9 @@ void SignatureExpansion::expandExternalSignatureTypes() {
     auto clangTy = IGM.getClangType(param, FnType);
+    if (clangTy->isVoidType()) {
+      continue;
+    }
```
This probably deserves a comment explaining that you're smuggling ignored parameters through lowering this way.

zoecarver replied:
An alternative (maybe better?) approach would be skipping metatypes two lines above before we ever even convert them to void. Then we could remove support for thin metatypes in `GenClangType` too.

rjmccall replied:
Why not just drop this argument from the SILFunctionType instead of working around it in IRGen?  I'm positive we have support for that in SIL type lowering; maybe it needs to be extended to constructors, but that shouldn't be hard.

zoecarver replied:
Yeah, I just thinking about the possibility of doing this. I wonder how many things depend on the constructor taking a metatype. I bet you're right, we could just remove this parameter from the SILFunction. It's going to get removed at some point anyway.

zoecarver replied:
I've update this patch to skip the self metatype parameter. Let me know what you think.

rjmccall reviewed (commented):
If we're going to emit a thunk for the constructor anyway, why do all the extra tricks to handle it having been lowered as a `@convention(c)` type?   We could just lower it with a native type (but with the arguments bridged, I guess) and then always introduce the thunk.

martinboehme committed:
Respond to review comments.

martinboehme committed:
Another response to review comments.

martinboehme committed:
Add a test that Objective-C types passed to a C++ constructor are
bridged correctly.

martinboehme committed:
Add SILGen and IRGen tests for passing Objective-C types to C++
constructors.

martinboehme committed:
Rename constructors-sil.swift to constructors-silgen.swift.

martinboehme commented:
@swift-ci Please test Windows

martinboehme commented:
@rjmccall Wanted to check in and see if you've had a chance to look at my changes and comments. Understand you may be busy with WWDC, just wanted to make sure this hadn't slipped off your radar. Thanks!

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3436,12 +3441,19 @@ namespace {
+      if (hasZeroInitializableStorage && !cxxRecordDecl) {
+        // Add default constructor for the struct if compiling in C mode.
+        // If we're compiling for C++, we'll import the C++ default constructor
+        // (if there is one), so we don't need to synthesize one here.
```
What do you think about adding an assertion in `synthesizeStructDefaultConstructorBody` to make sure we don't try to synthesize a (likely incorrect) constructor using `Builtin.zeroInitializer`?

rjmccall replied:
That seems good.  I wouldn't want us to accidentally miscompile in some weird case.

zoecarver replied:
Agreed.

MForster commented:
@rjmccall, could you please take another look at this pull request? A lot of our C++ interop work is blocked on this PR. It would be good if we could move forward, now that WWDC is over.

zoecarver commented:
It would be really great to move this along. There are a few things waiting on it.

hlopko commented:
@rjmccall friendly ping :)

zoecarver commented on test/Interop/Cxx/class/Inputs/constructors.h:
```diff
@@ -0,0 +1,39 @@
+  int x;
+};
+
+struct ImplicitDefaultConstructor {
```
We should also have a non-trivial version of `ImplicitDefaultConstructor`. See [the issue in #32591](https://github.com/apple/swift/pull/32591#issuecomment-650841400).

zoecarver replied:
`StructWithCopyConstructorAndValue` accomplishes this.

zoecarver committed:
[cxx-interop] Fix missing APIs and tests after rebase.

 * Update tests that relied on old behavior.
 * Use mangleCXXName instead of mangleCXXCtor.
 * Call VisitCXXRecordDecl not VisitRecordDecl from
 VisitClassTemplateSpecializationDecl. This allows template constructors
 to be imported and called correctly.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test linux platform

zoecarver commented:
@swift-ci please test Windows platform

zoecarver commented:
A change in the past day or two made it so that `type->hasRepresentation()` is no longer true in `visitMetatypeType`. Fixing...

zoecarver committed:
[cxx-interop] Revert changes to ClangTypeConverter::visitMetatypeType.

The old implementation should be sufficient. The C++ constructors we
import should get their types from the imported decl, not a double
conversion to a SILFunction type and then back again.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please test windows platform

zoecarver committed:
[cxx-interop] Skip void types in emitCXXConstructorThunkIfNeeded.

emitCXXConstructorThunkIfNeeded receives a SIL function that may contain
metatypes. If so, they will be converted to void and we need to skip
them when emitting the thunk's args.

zoecarver commented:
Let's see if that fixes the assertions on windows.

@swift-ci please test windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test windows platform.

zoecarver committed:
[cxx-interop] Update ABIArgInfo::Indirect case in expandExternalSignatureTypes to handle thin metatype parameters.

Add second list of SILParameterInfos that excludes thin metatypes so
that the "Indirect" argument path below will select the correct
parameter info.

zoecarver commented:
Unfortunately, right now the windows crashes are a bit of a black box for me. I can only guess at what the issue may be. Hopefully, this most recent commit will fix it. Otherwise, I may have to set up a windows machine to debug this locally.

zoecarver commented:
@swift-ci please test windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

rjmccall commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3434,12 +3443,19 @@ namespace {
+      bool hasUserDeclaredConstructor =
+          cxxRecordDecl && cxxRecordDecl->hasUserDeclaredConstructor();
+      if (hasReferenceableFields && hasMemberwiseInitializer &&
+          !hasUserDeclaredConstructor) {
```
This is creating a memberwise initializer, right?  Or at least a partial one for the members we can import?  For C++, this is strongly analogous to aggregate initialization, and the condition we should be using is `isAggregate()`.  If that's wrong, we at least need to be explaining why in a comment.

zoecarver replied:
Makes sense to me. Fixed.

zoecarver replied:
Actually, there is a slight difference. If a struct has a deleted constructor, it's still an aggregate, but (with the `hasUserDeclaredConstructor` condition above), we wouldn't generate a "memberwise initializer." C++ [allows aggregate initialization even if there's an explicitly deleted constructor](https://godbolt.org/z/v45sWs), so I think this is the correct behavior. But I figured it would be a good to point it out anyway.

rjmccall commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3470,8 +3486,9 @@ namespace {
-        result->setIsCxxNonTrivial(!cxxRecordDecl->isTriviallyCopyable());
+      if (cxxRecordDecl) {
+        result->setIsCxxNonTrivial(
+            !cxxRecordDecl->isTriviallyCopyable());
```
This is unrelated to this PR, but are we checking properly for non-copyable types?

zoecarver replied:
I'm not 100% sure. We'll need to look into this more deeply in #32378.

rjmccall commented on lib/ClangImporter/ImportName.cpp:
```diff
@@ -1600,6 +1610,15 @@ ImportedName NameImporter::importNameImpl(const clang::NamedDecl *D,
+    isFunction = true;
+    result.info.initKind = CtorInitializerKind::Designated;
+    baseName = "init";
+    if (auto ctor = dyn_cast<clang::CXXConstructorDecl>(D)) {
```
I don't think this cast can fail for declarations with a CXXConstructorName name.

rjmccall commented on lib/ClangImporter/ImportName.cpp:
```diff
@@ -1366,6 +1366,16 @@ static bool suppressFactoryMethodAsInit(const clang::ObjCMethodDecl *method,
 
+static void
+addEmptyArgNamesForCxxFunc(const clang::FunctionDecl *funcDecl,
+                           SmallVectorImpl<StringRef> &argumentNames) {
```
Our usual naming style would be to spell out Function.  It's not specific to C++ at all, so please drop CXX from the name  — you extracted this code from a function that was applying it to C functions!

zoecarver replied:
I concur. Renamed to `addEmptyArgNamesForClangFunction`.

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1228,6 +1257,18 @@ void SignatureExpansion::expandExternalSignatureTypes() {
+  if (formalIndirectResult) {
+    // If the result is a formal indirect result in SIL, that means that the
+    // Clang function has an explicit output parameter (e.g. it's a C++
+    // constructor). This means:
```
This is a foreign function convention; why aren't we taking the attributes mostly from Clang?

zoecarver replied:
Hmm. I agree it would make sense that clang should have the correct attributes on this (clang) function. It doesn't look like we override them anywhere and the generated IR doesn't seem to have a `nocapture` or `sret` on any of the constructor functions. The reasoning seems solid but I'm not sure why @martinboehme thought clang might not do the correct thing.

Removing this change seems fine to me. If I understand correctly, that's what you're suggesting as well, John. Does anyone see a reason this change should stay?

hlopko replied:
Hmm maybe @gribozavr remembers/knows. FWIW, if tests pass, let's clean it up.

zoecarver replied:
OK. I'm going to remove it.

zoecarver commented:
@swift-ci please test windows platform.

zoecarver commented:
@swift-ci please test windows platform.

zoecarver commented:
@swift-ci please test windows platform.

zoecarver committed:
[cxx-interop] Skip metatypes when lowering C++ constructor SIL function type.

When lowering a C++ constructor's function type to a SIL function type,
skip over the "self" metatype parameter.

zoecarver committed:
[cxx-interop] Small fixes and cleanup based on review.

 * Check isAggregate instead of hasUserDeclaredConstructor.
 * Rename addEmptyArgNamesForCxxFunc ->  addEmptyArgNamesForClangFunction.
 * Other minor fixes and cleanups.

zoecarver committed:
[cxx-interop] Fix patch formatting with clang-format.

Fix all formatting of the changes made by this patch.

zoecarver commented:
@swift-ci please test windows platform.

zoecarver commented:
@swift-ci please test windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
Looks like an unrelated stdlib issue.

@swift-ci please test Windows platform.

zoecarver committed:
[cxx-interop] Generate memberwise initializers for non-C++ types.

Previously, when we checked for `!hasUserDeclaredConstructor` rather than
`isAggregate`, we would always generate memberwise initializers for C
record types. This updates the isAggregate to be true for non-C++ types
so we will generate memberwise initializers for them as well.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver committed:
[cxx-interop] Bail if trying to convert call result to void.

Don't try to coerce the result of a call to void. If the "expected"
result type is void, just bail.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented on lib/ClangImporter/ImportName.cpp:
```diff
@@ -1600,6 +1609,14 @@ ImportedName NameImporter::importNameImpl(const clang::NamedDecl *D,
+    isFunction = true;
+    result.info.initKind = CtorInitializerKind::Designated;
+    baseName = "init";
+    if (auto ctor = dyn_cast<clang::CXXConstructorDecl>(D))
```
Re John's earlier comment: it would appear that _on windows only_ `CXXConstructorName` does not necessarily indicate `D` is a `CXXConstructorDecl`. I'll investigate this further because it seems a bit odd.

zoecarver committed:
[cxx-interop] Fix Windows tests for non-trivial C++ types.

* Replace `??1` with `??0` prefix in SILGen tests.
* Replace `call void <constructor>` with `call {{.*}} <constructor>`
because Windows ABI sometimes returns a pointer from the constructor for
non-trivial types.
* Make `noalias` an optional check.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Linux platform.
@swift-ci please test OS X platform.

zoecarver commented:
@swift-ci please test Linux platform.

zoecarver commented:
@swift-ci please test Linux platform.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3491,6 +3505,37 @@ namespace {
+      if (decl->getDefinition() && !decl->isBeingDefined() &&
+          !decl->isDependentContext() &&
+          decl->needsImplicitDefaultConstructor()) {
+        // Casting away const here should be OK because
```
I'm not sure we need this comment at all but if it stays it should say, "Casting away const here is be OK because..."

zoecarver commented:
Finally, the builds are green! There are a few minor review comments I need to address, but I think this patch is almost good to go. 

@rjmccall, @hlopko, and @gribozavr thank you for all your help reviewing this patch. Do you have any more comments?

lattner commented:
Super exciting @zoecarver !!!

hlopko commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3855,27 +3902,48 @@ namespace {
+        FuncDecl *func =
+            createFuncOrAccessor(Impl.SwiftContext, loc, accessorInfo, name,
+                                 nameLoc, bodyParams, resultTy,
+                                 /*async*/ false, /*throws*/ false, dc, decl);
```
Nit: `/*async=*/,.../*throws=*/` (the = sign)

hlopko reviewed (approved):
Thanks for pushing this forward!

zoecarver committed:
[cxx-interop] Remove logic around applying attributes.

Removes the logic around applying attributes to C++ constructor's
indirect results.

Also fixes some commenting.

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1417,6 +1429,18 @@ void SignatureExpansion::expandExternalSignatureTypes() {
+    // - The Clang function might capture the pointer, so don't specify
+    //   `nocapture`.
+    addIndirectResultAttributes(IGM, Attrs, 0, /* allowSRet = */ false,
+                                /* noCapture = */ false);
```
So, technically, the "i.e." doesn't hold here — we can have a semantically-indirect result that really isn't allowed to be captured this way.  That would only happen with an aggregate that wasn't a non-trivial C++ type, though: for example, a struct with an ObjC `__weak` field or a field with an address-diversified `__ptrauth` type.  So there are other reasons to maybe just let Clang handle these attributes.

rjmccall replied:
Also, general question: have you tested this on an ARM target, e.g. Apple ARM64?  Because the ARM C++ ABI actually gives constructors and destructors a return value (they return `this`), so we need to make sure we don't have unexpected mismatches.

zoecarver replied:
> Also, general question: have you tested this on an ARM target, e.g. Apple ARM64?

I think the bots run tests on a simulator so that should test on ARM64. We also have some tests that run on multiple targets no matter the local environment. (Semi-unrelated: in order to run the swift compiler for multiple targets, these tests use `-parse-as-library` which can cause odd behavior so we may want to move away from that in the future.)

> Because the ARM C++ ABI actually gives constructors and destructors a return value (they return this), so we need to make sure we don't have unexpected mismatches.

IIRC the Windows ABI also returns a pointer from constructors of non-trivial types. I had to update a few tests to accommodate that recently.

zoecarver replied:
After removing this, I discovered another reason I think keeping `addIndirectResultAttributes` here is incorrect: `addIndirectResultAttributes` will add the `noalias` attribute to the output pointer argument which I am pretty sure is not a safe assumption to make. There's no guarantee that the (user-provided) constructor will not write to a global variable, for example.

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -2789,6 +2862,11 @@ llvm::Function *IRGenModule::getAddrOfSILFunction(
+
+    if (auto ctor = dyn_cast<clang::CXXConstructorDecl>(clangDecl)) {
+      clangAddr =
+          emitCXXConstructorThunkIfNeeded(*this, f, ctor, entity, clangAddr);
```
Hmm, so we're always going through the thunk?  Alright, I guess this is okay.

zoecarver replied:
`emitCXXConstructorThunkIfNeeded` will exit early (returning `clangAddr`) if the thunk isn't needed.

rjmccall replied:
Sure, that's not quite what I meant but it's okay.

zoecarver replied:
OK. If you want me to change something here, just let me know.

rjmccall replied:
No, it's fine, I don't think there's any reason for us to do extra work to try to always do a direct call.

zoecarver replied:
I agree, at least for now. Maybe in the future. I think it would be doable.

rjmccall reviewed (commented):
Looking really good.

zoecarver commented on lib/IRGen/GenCall.cpp:
```diff
@@ -338,10 +338,12 @@ llvm::CallingConv::ID irgen::expandCallingConv(IRGenModule &IGM,
                                         llvm::AttributeList &attrs,
-                                        unsigned paramIndex, bool allowSRet) {
+                                        unsigned paramIndex, bool allowSRet,
+                                        bool noCapture = true) {
```
Also, this adds the `nocapture` attribute in _all the places_ `addIndirectResultAttributes` is used, which I'm not sure is always correct. I'm surprised more tests didn't fail because of this.

rjmccall replied:
Yeah, I'm not quite sure.  Swift doesn't currently expose the ability to rely on an escaped pointer into a value, at least in user code — it can happen in hidden ways, though, like in weak references.

zoecarver commented on test/Interop/Cxx/class/constructors-irgen.swift:
```diff
@@ -0,0 +1,94 @@
+struct UnsafePointer<T> { }
+struct UnsafeMutablePointer<T> { }
+
+public func createHasVirtualBase() -> HasVirtualBase {
```
Should I remove the below comments now that the patch no longer modifies any of the logic around applying attributes, or is the information still helpful?

rjmccall replied:
I'd drop them.

zoecarver replied:
SGTM.

zoecarver commented on test/Interop/Cxx/class/synthesized-initializers-silgen.swift:
```diff
@@ -1,37 +1,26 @@
 // RUN: %target-swift-frontend -I %S/Inputs -enable-cxx-interop -emit-silgen %s | %FileCheck %s
 
 import SynthesizedInitializers
+import Constructors
```
I'm going to merge this test with `constructors-silgen`. It doesn't make sense to have a "synthesized initializers" test when we never actually synthesize an initializer.

zoecarver committed:
[cxx-interop] Merge synthesized initializer tests into "constructors-silgen".

We no longer synthesize initializers for C++ types so this name no longer makes sense.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
Dropped 702689480ae3a9b018ceba5f5cca7de22ee54c3c. I'm going to double-check that this crashes before I launch into debugging it.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver committed:
[cxx-interop] Look through template decl to find constructor when importing the function name.

Sometimes, on windows, we get a function template wrapping the
constructor decl. In this case, look through the function template to
find the constructor decl.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
It looks like (on Windows only?) it's possible for a function template to have a `CXXConstructorName` if it wraps a `CXXConstructorDecl`. I wonder if #33053 will fix this (I doubt it). But, if that doesn't work, I think a fairly simple fix would be to just get the constructor decl out of the function template if we find a function template in that branch.

zoecarver commented:
https://github.com/apple/swift/pull/33053
@swift-ci please test Windows platform

zoecarver commented:
Yeah, that didn't work. I'll try my other idea.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@rjmccall thank you again for the help reviewing. I really appreciate your insights. Any other comments? 

Hoping to land this in the next few days.

zoecarver commented:
Unless anyone has an objection, I'm going to land this patch on Wednesday.

lattner commented:
nice!

rjmccall commented:
> It looks like (on Windows only?) it's possible for a function template to have a `CXXConstructorName` if it wraps a `CXXConstructorDecl`.

Oh, yes, that's right.  I don't think that's Windows only — any constructor template should have that characteristic.

rjmccall reviewed (approved):
LGTM.  Thanks for seeing this through!  Very impressive work.

zoecarver commented:
Thanks, @rjmccall. Given that you've both approved this PR, I'm going to go ahead and land it now. 

Before it lands, I want to run the benchmarks, though. As much to test #32721 as anything else. Hopefully, this swift-ci incantation will work.

zoecarver commented:
https://github.com/apple/swift/pull/32721
@swift-ci please benchmark.

zoecarver commented:
Actually, thinking about it, that won't work because even if it does correctly merge the branches before benchmarking, the benchmark will be shown as "added" so, we won't be able to get any real data from it.

zoecarver commented:
@swift-ci please test and merge.

zoecarver commented:
@swift-ci please test and merge.

hlopko commented:
Big congrats to both Martin and Zoe, impressive work!
